vm/chef/cookbooks/mariadb/files/mariadb-utils

```diff
@@ -0,0 +1,323 @@
+
+    run_query "
+      GRANT REPLICATION SLAVE ON *.*
+      TO '${replication_user}'@'${replica_ip}';"
+
+      emit_msg "Replication user created."
+  done
+}
+
+#######################################
+# Configure root user authorization.
+# Arguments:
+#   Root user password.
+#   Enable root access from any host or not.
+#######################################
+function set_root_user() {
+  local -r root_password="$1"
+  local -r enable_root_anywhere="$2"
+
+  echo "Set root privileges..."
+  echo "Enable root from anywhere: ${enable_root_anywhere}"
+
+  # Set root password
+  run_query "
+    SET PASSWORD FOR 'root'@'localhost' = PASSWORD('${root_password}');
+    FLUSH PRIVILEGES;"
+
+  # Grant privilege to root user connect from anywhere
+  if [[ "${enable_root_anywhere}" == "True" ]]; then
+    run_query "CREATE USER 'root'@'%' IDENTIFIED BY '${root_password}';"
+    run_query "GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' WITH GRANT OPTION";
+  fi
+}
+
+#######################################
+# Create backup from all available databases
+# and replication information.
+#######################################
+function create_backup() {
+  run_query "CREATE DATABASE ${DATABASE_NAME};"
+  mkdir -p "${backup_folder}"
+  mysqldump --all-databases --user root --master-data > "${backup_filename}"
+}
+
+#######################################
+# Await for replicas set up ssh access
+# to the replication user
+# Arguments:
+#  Replicas list separated by spaces.
+#######################################
+function await_for_replicas_ssh() {
+  local -r replicas="$1"
+
+  for replica in ${replicas[@]}; do
+    emit_msg "Checking SSH on ${replica}..."
+    until ssh -o StrictHostKeyChecking=no ${replication_user}@${replica} -- cat /dev/null; do
+      sleep 1
+    done
+    emit_msg "Connected to ${replica} SSH with success."
+  done
+}
+
+#######################################
+# Send backup via SSH to all replicas available
+# Arguments:
+#  Replicas list separated by spaces.
+#######################################
+function send_backup_to_replicas() {
+  local -r replicas="$1"
+
+  emit_msg "Sending backup to replicas..."
+  for replica in ${replicas[@]}; do
+    echo "Sending backup to ${replica}..."
+    scp -o StrictHostKeyChecking=no "${backup_filename}" \
+      "${replication_user}@${replica}:${backup_folder}"
+    echo "Backup sent to ${replica}."
+  done
+}
+
+#######################################
+# Setup a master MariaDB node.
+# Arguments:
+#   Replicas list separated by spaces.
+#   root user password.
+#   replication user password.
+#   Enable root access from any host or not.
+#######################################
+function setup_master() {
+  local -r replicas="$1"
+  local -r root_password="$2"
+  local -r replication_password="$3"
+  local -r enable_root_anywhere="$4"
+
+  apply_master_replication_config
+
+  # Start and await service to be up
+  systemctl start mariadb
+  until mysqladmin ping; do sleep 1; done
```

Comment:
maybe there should be fail in case of timeout?

Reply:
everywhere where you have waiting

Reply:
Done.